dashbuilder/dashbuilder-backend/dashbuilder-services/src/main/java/org/dashbuilder/DashbuilderMigrationService.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dashbuilder;
+
+import java.util.HashMap;
+import java.util.function.Predicate;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.uberfire.backend.server.spaces.SpacesAPIImpl;
+import org.uberfire.commons.services.cdi.Startup;
+import org.uberfire.io.IOService;
+import org.uberfire.java.nio.file.Path;
+import org.uberfire.java.nio.file.FileSystem;
+import org.uberfire.java.nio.file.SimpleFileVisitor;
+import org.uberfire.java.nio.file.attribute.BasicFileAttributes;
+import org.uberfire.java.nio.file.FileVisitResult;
+import org.uberfire.java.nio.file.Files;
+import org.uberfire.java.nio.IOException;
+import org.uberfire.spaces.SpacesAPI;
+
+import static org.kie.soup.commons.validation.PortablePreconditions.checkNotNull;
+
+@ApplicationScoped
+@Startup
+public class DashbuilderMigrationService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DashbuilderMigrationService.class);
+
+    private IOService ioService;
+    private FileSystem oldDatasetsFS;
+    private FileSystem datasetsFS;
+    private FileSystem pluginsFS;
+    private FileSystem perspectivesFS;
+    private FileSystem navigationFS;
+
+    public DashbuilderMigrationService() {
+
+    }
+
+    @Inject
+    public DashbuilderMigrationService(
+            final @Named("ioStrategy") IOService ioService,
+            final @Named("datasetsFS") FileSystem datasetsFS,
+            final @Named("pluginsFS") FileSystem pluginsFS,
+            final @Named("perspectivesFS") FileSystem perspectivesFS,
+            final @Named("navigationFS") FileSystem navigationFS) {
+
+        this.ioService = ioService;
+
+        this.oldDatasetsFS = ioService.getOrNewFileSystem(
+            new SpacesAPIImpl().resolveFileSystemURI(
+                SpacesAPI.Scheme.DEFAULT,
+                SpacesAPI.DEFAULT_SPACE,
+                "datasets"),
+            new HashMap<String, Object>());
+
+        this.datasetsFS = datasetsFS;
+        this.pluginsFS = pluginsFS;
+        this.perspectivesFS = perspectivesFS;
+        this.navigationFS = navigationFS;
+    }
+
+    @PostConstruct
+    private void init() {
+        migrateDatasets(oldDatasetsFS, datasetsFS);
```

Comment:
My suggestion here is to first, instead of creating the datasets if doesn't exist (oldDatasetsFS), just do a getFilesystem, and if oldDatasetsFS is null doesn't run migration at all (this makes the startup a little faster).
Also, in the end of the migration process, you should delete the oldDatasetsFS filesystem.

Reply:
Using `getFileSystem` just doesn't work due to that CDI bug with on `@Startup` with FileSystem
In order to make it work, I used the `try/catch` block with `newFileSystem/getFileSystem`.
I ran several tests and its working fine. I could not find a way to bypass that bug

Reply:
I changed my old commit to delete `git://system/datasets` after Migration.
I am keeping `git://system/plugins` because it contains important data that cannot be deleted, but I am deleting all files related to perspectives and navigation after Migration.